samples/net/sockets/echo/src/socket_echo.c

```diff
@@ -51,15 +52,30 @@ int main(void)
 		printf("Connection #%d from %s\n", counter++, addr_str);
 
 		while (1) {
-			char buf[128];
+			char buf[128], *p;
 			int len = recv(client, buf, sizeof(buf), 0);
+			int out_len;
 
-			if (len == 0) {
+			if (len <= 0) {
+				if (len < 0) {
+					printf("error: recv: %d\n", errno);
+				}
 				break;
 			}
-			send(client, buf, len, 0);
+
```

Comment:
Looks fine to me.

Also, it seems like a good opportunity to hide this send-loop under the new socket/stream abstraction, so each user of send() doesn't have to replicate this code in every socket-based program.

Reply:
Such a helper function can be added to stream API. What signature do you think will be good for it?

Reply:
I recommend actually placing it under the new abstraction API (currently proposed as  zstream_write(stream, p, len), rather than adding a helper function.

Reply:
Well, POSIX API doesn't have "superfluous" or "unnecessary" functionality. The reason there can be short writes is because there would be a blocking operation otherwise, or extra resource usage. Repeating the operation nonetheless is just one strategy of dealing with it, and not very efficient one. A more efficient (and complex) strategy is to switch to doing something else in this case (async I/O scheduling). All those opportunities are encoded in POSIX semantics, and the task of streams API is to preserve it faithfully. So, zstream_write() would behave exactly the same as write()/send().

We can add a helper function, e.g. zstream_writeall(), and either designate it as usuable only for a particular usecase (e.g. blocking writes), or increase its interface complexity trying to cover all POSIX cases (still with unlikely expectations that it will be used for something else but blocking operations).

E.g., a no-information-loss prototype would be:

~~~
ssize_t zstream_writeall(zstream s, const char *buf, size_t len, size_t *written);
~~~

with its peculiarities, like: number of bytes written will be returned by 2 means: via `size_t *written` out-param (if it's not NULL apparently), and as return value - unless it's error. (The whole purpose of `size_t *written` is to communicate back number of bytes written before an error occurred (errors can be recoverable and non-recoverable)).

We can only appreciate how much of semantics POSIX packs into pretty simple and unambiguous signatures ;-).

Reply:
OK, sounds reasonable.  The helper function is not really critical.

Reply:
I've added this function, to not forget about it, to the latest https://github.com/zephyrproject-rtos/zephyr/pull/5985 version (https://github.com/zephyrproject-rtos/zephyr/pull/5985/commits/762e71dee0399e586c5b630b155185f8c52f1be0 until rebased). Didn't update any samples to use it though, will do that later.